File: day04/high_entropy.pl
#!/usr/bin/env perl
use strict;
use warnings;
use feature ":5.24";
use experimental 'signatures';

my $cnt = 0;
my $cnt2 = 0;
while (<>) {
    chomp;
    $cnt++ if is_valid($_);
    $cnt2++ if is_valid2($_);
}

say "result1: $cnt";
say "result2: $cnt2";

sub is_valid($s) {
    my %h;
    my @a = split / /, $s;
    for (@a) {
        return 0 if defined $h{$_};
        $h{$_} = 1;
    }

    return 1;
}

sub is_valid2($s) {
    my %h;
    my @a = map {join "", sort split //} split / /, $s;
    for (@a) {
        return 0 if defined $h{$_};
        $h{$_} = 1;
    }

    return 1;
}

File: day04/high_entropy.py
#!/usr/bin/env python3
from sys import argv

def is_valid(line):
    s = set()
    for pw in line.split():
        if pw in s:
            return 0
        else:
            s.add(pw)
            
    return 1
    
def is_valid2(line):
    s = set()
    keys = [''.join(sorted(list(pw))) for pw in line.split()]
    for key in keys:
        if key in s:
            return 0
        else:
            s.add(key)
            
    return 1
    

filename = argv[1]
sum1 = 0
sum2 = 0

with open(filename) as f:
    for line in f:
        if is_valid(line):
            sum1 += 1
        if is_valid2(line):
            sum2 += 1

print("result1:", sum1)
print("result2:", sum2)
